content/blog/2022-04-07-compiling-black-with-mypyc-part-2.md

```diff
@@ -0,0 +1,537 @@
+CPyL4: ;
+    cpy_r_r6 = NULL;
+    return cpy_r_r6;
+}
+```
+
+If we mark the `SCALE` variable as Final, mypyc will notice that and inline the value.
+
+```python3 {hl_lines=[3]}
+from typing import Final
+
+SCALE: Final = 5
+
+def calculate_y(x):
+    return x * SCALE
+```
+
+Just look at how much shorter this is!
+
+```c {linenos=table, hl_lines=[6]}
+PyObject *CPyDef_calculate_y(PyObject *cpy_r_x) {
+    PyObject *cpy_r_r0;
+    PyObject *cpy_r_r1;
+    PyObject *cpy_r_r2;
+CPyL0: ;
+    cpy_r_r0 = CPyTagged_StealAsObject(10);
+    cpy_r_r1 = PyNumber_Multiply(cpy_r_x, cpy_r_r0);
+    CPy_DECREF(cpy_r_r0);
+    if (unlikely(cpy_r_r1 == NULL)) {
+        CPy_AddTraceback("final.py", "calculate_y", 6, CPyStatic_globals);
+        goto CPyL2;
+    }
+CPyL1: ;
+    return cpy_r_r1;
+CPyL2: ;
+    cpy_r_r2 = NULL;
+    return cpy_r_r2;
+}
+```
+
+If the value can't be inlined, say because it involves a function call then mypyc will
+just replace the globals dictionary lookup and instead use a C static (which is still
+faster).
+
+Applying this optimization to Black yields changes like the following (`black.comments`
+and `black.parsing` respectively):
+
+```diff
+@@ -12,11 +18,10 @@ from black.nodes import STANDALONE_COMMENT, WHITESPACE
+ # types
+ LN = Union[Leaf, Node]
+
+-
+-FMT_OFF = {"# fmt: off", "# fmt:off", "# yapf: disable"}
+-FMT_SKIP = {"# fmt: skip", "# fmt:skip"}
+-FMT_PASS = {*FMT_OFF, *FMT_SKIP}
+-FMT_ON = {"# fmt: on", "# fmt:on", "# yapf: enable"}
++FMT_OFF: Final = {"# fmt: off", "# fmt:off", "# yapf: disable"}
++FMT_SKIP: Final = {"# fmt: skip", "# fmt:skip"}
++FMT_PASS: Final = {*FMT_OFF, *FMT_SKIP}
++FMT_ON: Final = {"# fmt: on", "# fmt:on", "# yapf: enable"}
+
+
+ @dataclass
+```
+
+```diff
+@@ -36,6 +36,11 @@ except ImportError:
+     else:
+         ast3 = ast27 = ast
+
++if sys.version_info >= (3, 8):
++    TYPE_IGNORE_CLASSES: Final = (ast3.TypeIgnore, ast27.TypeIgnore, ast.TypeIgnore)
++else:
++    TYPE_IGNORE_CLASSES: Final = (ast3.TypeIgnore, ast27.TypeIgnore)
++
+
+ class InvalidInput(ValueError):
+     """Raised when input source code fails all parse attempts."""
+@@ -160,10 +165,7 @@ def stringify_ast(
+
+     for field in sorted(node._fields):  # noqa: F402
+         # TypeIgnore has only one field 'lineno' which breaks this comparison
+-        type_ignore_classes: Tuple[Type, ...] = (ast3.TypeIgnore, ast27.TypeIgnore)
+-        if sys.version_info >= (3, 8):
+-            type_ignore_classes += (ast.TypeIgnore,)
+-        if isinstance(node, type_ignore_classes):
++        if isinstance(node, TYPE_IGNORE_CLASSES):
+             break
+
+         try:
+```
+
+In total, this was definitely the most common optimization I applied throughout this whole
+journey. It's simple but effective!
+
+#### Taking advantage of early binding
+
+This one is really making a deal with mypyc. You give it static code and it gives you fast
+code in return ... by resolving called functions and names at compile time. I already
```

Comment:
Note to self, I really hate this phrasing the more I read it. Just get rid of it.